File: 234-palindrome-linked-list/palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:

    ListNode* reverse(ListNode* head)
    {
        ListNode* prev=NULL;
        ListNode* curr=head;
        while(curr)
        {
            ListNode* nextNode=curr->next;
            curr->next=prev;
            prev=curr;
            curr=nextNode;
        }
        return prev;
    }

    bool isPalindrome(ListNode* head) {
        if(!head || !head->next)
            return 1;
        ListNode* slow=head;
        ListNode* fast=head;
        while(fast->next && fast->next->next)
        {
            slow=slow->next;
            fast=fast->next->next;
        }
        ListNode* newHead=reverse(slow->next);
        ListNode* fpart=head;
        ListNode* spart=newHead; 
        while(spart)
        {
            if(fpart->val!=spart->val)
            {
                reverse(newHead);
                return 0;
            }
            fpart=fpart->next;
            spart=spart->next;
        }
        reverse(newHead);
        return 1;
    }
};

File: 234-palindrome-linked-list/palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {

    ListNode reverse(ListNode head)
    {
        if(head==null || head.next==null)
            return head;
        ListNode prev=null;
        ListNode curr=head;
        while(curr!=null)
        {
            ListNode nextNode=curr.next;
            curr.next=prev;
            prev=curr;
            curr=nextNode;
        }
        return prev;
    }

    public boolean isPalindrome(ListNode head) {
        if(head==null || head.next==null)
            return true;
        ListNode slow=head,fast=head;
        while(fast.next!=null && fast.next.next!=null)//O(n/2)
        {
            slow=slow.next;
            fast=fast.next.next;
        }
        ListNode newHead=reverse(slow.next);//O(n/2)
        // reverse(newHead);
        ListNode left=head,right=newHead;
        while(right!=null)//O(n/2)
        {
            if(right.val!=left.val)
            {
                reverse(newHead);
                return false;
            }
            left=left.next;
            right=right.next;
        }
        reverse(newHead);//O(n/2)
        return true;
    }
}